File: c++/правильная скобочная последовательность .cpp
#include <iostream>
#include <vector>
#include <set>

using namespace std;

int length;
void copy(int m)
{
    length = m;
}

bool true_queele(vector<char> data)
{
    set<char> open = {'(', '['};
    vector<char> stack;

    for (int i = 0; i < data.size(); i++)
    {
        if (open.find(data[i]) != open.end())
        {
            stack.push_back(data[i]);
        }
        else
        {
            if (stack.size() == 0)
            {
                return false;
            }
            else if (data[i] == ')' && stack[stack.size() - 1] == '[')
            {
                return false;
            }
            else if (data[i] == ']' && stack[stack.size() - 1] == '(')
            {
                return false;
            }
            else
            {
                stack.pop_back();
            }
        }
    }
    return true;
}

void generate_permutations(vector<char> n, int m, vector<char> prefix, int c_open)
{

    if (c_open > length / 2)
    {
        return;
    }
    if (m == 0)
    {
        for (auto i : prefix)
        {
            cout << i;
        }
        cout << '\n';
        return;
    }
    for (auto digit : n)
    {
        prefix.push_back(digit);
        if (digit == '(' || digit == '[')
        {
            c_open++;
        }
        if (!true_queele(prefix))
        {
            prefix.pop_back();
            continue;
        }
        generate_permutations(n, m - 1, prefix, c_open);
        if (digit == '(' || digit == '[')
        {
            c_open--;
        }
        prefix.pop_back();
    }
}

int main()
{
    vector<char> n = {'(', '[', ']', ')'};
    int m;
    cin >> m;
    copy(m);
    vector<char> prefix;
    int c_open = 0;
    if (m % 2 == 1)
    {
        cout << '\n';
    }
    else
    {
        generate_permutations(n, m, prefix, c_open);
    }
}

File: Python/правильная скобочная последовательность.py
m = int(input())
length = m

def true_queele2(data):
    open = {'(', '['}
    # c = {i: 0 for i in ['(', '[']}
    stack = []
    for i in range(len(data)):
        if data[i] in open:
            stack.append(data[i])
            # c[data[i]] += 1
        else:
            if len(stack) == 0:
                return False
            # elif c['('] + c['['] > m // 2:
            #     return False
            elif data[i] == ')' and stack[-1] == '[':
                return False
            elif data[i] == ']' and stack[-1] == '(':
                return False
            else:
                stack.pop(-1)
    return True

def generate_permutations(n, m=2, prefix=None, c_open=0, c_close=0):
    if c_open > length // 2:
        return
    if c_close > length // 2:
        return
    prefix = prefix or []
    if m == 0:
        print(''.join(prefix))
        return
    for digit in n:
        prefix.append(digit)
        if digit == '(' or digit == '[':
            c_open += 1
        if digit == ']' or digit == ')':
            c_close += 1
        if true_queele2(prefix) == False:
            if digit == ']' or digit == ')':
                c_close -= 1
            prefix.pop()
            continue
        generate_permutations(n, m-1, prefix, c_open, c_close)
        if digit == '(' or digit == '[':
            c_open -= 1
        if digit == ']' or digit == ')':
            c_close -= 1
        prefix.pop()

generate_permutations(n=['(', '[', ']', ')'], m=m)